Source/WebKit/Platform/cocoa/WKMaterialHostingSupport.swift

```diff
@@ -0,0 +1,162 @@
+// 1. Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+
+#if !os(visionOS) && canImport(SwiftUI, _version: 7.0.27)
+
+internal import WebKit_Internal
+
+#if canImport(UIKit)
+@_spi(Private) @_spi(ForUIKitOnly) internal import SwiftUI
+#else
+@_spi(Private) @_spi(ForAppKitOnly) internal import SwiftUI
+#endif
+
+#if canImport(UIKit)
+
+private struct MaterialHostingContentViewWrapper: UIViewRepresentable {
+    let contentView: UIView
+
+    func makeUIView(context: Context) -> UIView {
+        contentView
+    }
+
+    func updateUIView(_ uiView: UIView, context: Context) {
+    }
+}
+
+#endif
+
+protocol MaterialHostingProvider {
+    associatedtype Source
+
+    associatedtype Body: View
+
+    @ViewBuilder
+    static func view(for source: Source) -> Body
+}
+
+private struct LayerBackedMaterialHostingProvider: MaterialHostingProvider {
+    static func view(for source: CALayer) -> some View {
+        _CALayerView(type: CALayer.self) { layer in
+            layer.addSublayer(source)
+        }
+    }
+}
+
+#if canImport(UIKit)
+
+private struct ViewBackedMaterialHostingProvider: MaterialHostingProvider {
+    static func view(for source: UIView) -> some View {
+        MaterialHostingContentViewWrapper(contentView: source)
+    }
+}
+
+#endif
+
+private struct MaterialHostingView<P: MaterialHostingProvider>: View {
+    private let content: P.Source
+    private let useMaterialEffect: Bool
+    private let cornerRadius: CGFloat
+
+    init(content: P.Source, useMaterialEffect: Bool = false, cornerRadius: CGFloat = 0) {
+        self.content = content
+        self.useMaterialEffect = useMaterialEffect
+        self.cornerRadius = cornerRadius
+    }
+
+    var body: some View {
+        let view = P.view(for: content)
+
+        if useMaterialEffect {
+            view.materialEffect(Material._hostedBlurMaterial, in: .rect(cornerRadius: cornerRadius))
```

Comment:
This will probably be fine in the short term, but using `if` this way is a bit of an anti-pattern due to invalidating the view’s identity between branches. If `view` had internal state this could result in it flip-flopping or being lost when the condition changes. Without the definition for `materialEffect(_:in:)` it’s hard to say, but generally SwiftUI modifiers should provide you an inert form so that you can sink those conditions into the arguments, while “structurally” the modifiers remain unconditional.

I also kinda wonder if a `ViewModifier` might be more correct here, just given how the structure of this looks.

Reply:
Thanks for this, and the other feedback. I've filed https://bugs.webkit.org/show_bug.cgi?id=287735 to track addressing the feedback.